### PR TITLE
[WIP] feat(backup): Add functionality to create backup

### PR DESCRIPTION
**Pull request by payes**
This PR will enable users to take backup of openebs volumes.

A rest endpoint is exposed in m-apiserver through which backup can be taken:
```/latest/backup/```
When the backup api is hit, a snapshot is created with a randomly generated name.
If the snapshot command is successful, a healthy replica is selected and a backup event is triggered on it.

### CODE REVIEW

File: cmd/maya-apiserver/app/server/backup_endpoint_v1alpha1.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+		CasType:    backup.Spec.CasType,
+		Name:       backup.Name,
+	})
+	if err != nil {
+		return nil, CodedError(400, err.Error())
+	}
+
+	glog.Infof("Creating %s volume %q snapshot", backup.Spec.CasType, backup.Spec.VolumeName)
+
+	snap, err := snapOps.Create()
```
Comment (utkarshmani1997) [inline cmd/maya-apiserver/app/server/backup_endpoint_v1alpha1.go]: is this call synchronous or asynchrounous ?

File: cmd/maya-apiserver/app/server/backup_endpoint_v1alpha1.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	if len(strings.TrimSpace(backup.Spec.BackupIP)) == 0 {
+		return nil, CodedError(400, fmt.Sprintf("failed to create backup '%v': missing backupIP", backup.Name))
+	}
+
+	// namespace is expected
+	if len(strings.TrimSpace(backup.Namespace)) == 0 {
+		return nil, CodedError(400, fmt.Sprintf("failed to create backup '%v': missing namespace", backup.Name))
+	}
+
+	//TODO Create snapname randomly
```
Comment (utkarshmani1997) [inline cmd/maya-apiserver/app/server/backup_endpoint_v1alpha1.go]: in PR description, it is mentioned that snapshot name is autogenerated. If this is true then please remove the above comment.

File: cmd/cstor-pool-mgmt/controller/backup-controller/handler.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+func IsPendingStatus(csb *apis.CStorBackup) bool {
+	if string(csb.Status.Phase) == string(apis.CSBStatusPending) {
+		glog.Infof("csb pending: %v", string(csb.ObjectMeta.UID))
+		return true
+	}
+	glog.V(4).Infof("Not pending status: %v", string(csb.ObjectMeta.UID))
+	return false
+}
+
+// getVolumeReplicaResource returns object corresponding to the resource key
```
Comment (sonasingh46) [inline cmd/cstor-pool-mgmt/controller/backup-controller/handler.go]: nit :- (comment) getVolumeReplicaResource -->getCStorBackupResource

File: cmd/cstor-pool-mgmt/controller/backup-controller/handler.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	// Convert the key(namespace/name) string into a distinct name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil, nil
+	}
+
+	CStorBackupUpdated, err := c.clientset.OpenebsV1alpha1().CStorBackups(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		// The cStorPool resource may no longer exist, in which case we stop
```
Comment (sonasingh46) [inline cmd/cstor-pool-mgmt/controller/backup-controller/handler.go]: nit: Comment

Comment (sonasingh46) [reply cmd/cstor-pool-mgmt/controller/backup-controller/handler.go]: Similarly at other places

File: cmd/maya-apiserver/app/server/backup_endpoint_v1alpha1.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	switch req.Method {
+	case "POST":
+		return backupOp.create()
+	}
+	return nil, CodedError(405, ErrInvalidMethod)
+}
+
+// Create is http handler which handles backup-create request
+func (bOps *backupAPIOps) create() (interface{}, error) {
+	glog.Infof("Backup create request was received")
```
Comment (vishnuitta) [inline cmd/maya-apiserver/app/server/backup_endpoint_v1alpha1.go]: what is the default loglevel?

File: pkg/apis/openebs.io/v1alpha1/cstor_backup.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +resource:path=cstorbackups
+
+// CStorBackup describes a cstor volume resource created as custom resource
```
Comment (vishnuitta) [inline pkg/apis/openebs.io/v1alpha1/cstor_backup.go]: cstor volume resource?

File: cmd/maya-apiserver/app/server/backup_endpoint_v1alpha1.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+	backup := &v1alpha1.CStorBackup{}
+
+	err := decodeBody(bOps.req, backup)
+	if err != nil {
+		return nil, err
+	}
+
+	// namespace is expected
+	if len(strings.TrimSpace(backup.Namespace)) == 0 {
```
Comment (vishnuitta) [inline cmd/maya-apiserver/app/server/backup_endpoint_v1alpha1.go]: is Namespace member variable?

**Comment by AmitKumarDas** [comment]
Closing this since this has become stale

**Pull request closed without merging**